controllers/machine_controller_test.go

```diff
@@ -29,16 +28,238 @@ import (
 	"k8s.io/client-go/kubernetes/scheme"
 	"k8s.io/utils/pointer"
 	clusterv1 "sigs.k8s.io/cluster-api/api/v1alpha2"
+	"sigs.k8s.io/controller-runtime/pkg/client"
 	"sigs.k8s.io/controller-runtime/pkg/client/fake"
 	"sigs.k8s.io/controller-runtime/pkg/log"
 	"sigs.k8s.io/controller-runtime/pkg/reconcile"
 )
 
-var _ = Describe("Machine Reconciler", func() {
-	It("Should create a Machine", func() {
-		// TODO
-	})
-})
+func TestMachineFinalizer(t *testing.T) {
+	RegisterTestingT(t)
+	clusterCorrectMeta := &clusterv1.Cluster{
+		ObjectMeta: metav1.ObjectMeta{
+			Namespace: "default",
+			Name:      "valid-cluster",
+		},
+	}
+
+	machineValidCluster := &clusterv1.Machine{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      "machine1",
+			Namespace: "default",
+			Labels: map[string]string{
+				clusterv1.MachineClusterLabelName: "valid-cluster",
+			},
+		},
+	}
+
+	machineWithFinalizer := &clusterv1.Machine{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      "machine2",
+			Namespace: "default",
+			Labels: map[string]string{
+				clusterv1.MachineClusterLabelName: "valid-cluster",
+			},
+			Finalizers: []string{"some-other-finalizer"},
+		},
+	}
+
+	testCases := []struct {
+		name               string
+		request            reconcile.Request
+		m                  *clusterv1.Machine
+		expectedFinalizers []string
+	}{
+		{
+			name: "should add a machine finalizer to the machine if it doesn't have one",
+			request: reconcile.Request{
+				NamespacedName: types.NamespacedName{
+					Name:      machineValidCluster.Name,
+					Namespace: machineValidCluster.Namespace,
+				},
+			},
+			m:                  machineValidCluster,
+			expectedFinalizers: []string{clusterv1.MachineFinalizer},
+		},
+		{
+			name: "should append the machine finalizer to the machine if it already has a finalizer",
+			request: reconcile.Request{
+				NamespacedName: types.NamespacedName{
+					Name:      machineWithFinalizer.Name,
+					Namespace: machineWithFinalizer.Namespace,
+				},
+			},
+			m:                  machineWithFinalizer,
+			expectedFinalizers: []string{"some-other-finalizer", clusterv1.MachineFinalizer},
+		},
+	}
+
+	for _, tc := range testCases {
+		t.Run(tc.name, func(t *testing.T) {
+			clusterv1.AddToScheme(scheme.Scheme)
+			mr := &MachineReconciler{
+				Client: fake.NewFakeClientWithScheme(
+					scheme.Scheme,
+					clusterCorrectMeta,
+					machineValidCluster,
+					machineWithFinalizer,
+				),
+				Log: log.Log,
+			}
+
+			mr.Reconcile(tc.request)
+
+			key := client.ObjectKey{Namespace: tc.m.Namespace, Name: tc.m.Name}
+			var actual clusterv1.Machine
+			if len(tc.expectedFinalizers) > 0 {
+				Expect(mr.Client.Get(ctx, key, &actual)).ToNot(HaveOccurred())
+				Expect(actual.Finalizers).ToNot(BeEmpty())
+				Expect(actual.Finalizers).To(Equal(tc.expectedFinalizers))
+			} else {
+				Expect(actual.Finalizers).To(BeEmpty())
+			}
```

Comment:
@ncdc I've made the assertions a little more clearer by not re-using `tc.m` for the actual machine object. I still passed `m` into the test case so it can be used for the `key`.